### PR TITLE
[DO-829] Generate PDF Invoice document v0.1

### PR DESCRIPTION
**Pull request by Andrii17s**
Still not done (blocked by [DO-792] and lack of seller_config data), but may be reviewed.

### CODE REVIEW

File: payment_system/templates/payment_system/invoice_det_uk.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+{% block seller %}
+    <h1>Продавець: </h1><br />
+     <br />
+     <br />
+    <h1>Одержувач: </h1>
+{% endblock seller %}
+
+{% block title %}
+<p>Рахунок №{{company.id}} від {% load i18n %} {% language 'ukr' %} {{company.start_date|date:'d E Y'}} {% endlanguage %} p.</p>
```
Comment (roman-tiukh) [inline payment_system/templates/payment_system/invoice_det_uk.html:11]: move `load` to top

Comment (roman-tiukh) [inline payment_system/templates/payment_system/invoice_det_uk.html:11]: ukr? maybe uk?

File: payment_system/templates/payment_system/invoice_detail.html [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        <td></td>
+        <td>{{wpdv|floatformat:2}}</td>
+    </tr>
+    </tbody>
+</table>
+{% block summ %}
+<h2>Summ: {{summ}}</h2>
+{% endblock %}
+</body>
+</html>
```
Comment (roman-tiukh) [inline payment_system/templates/payment_system/invoice_detail.html:48]: blank line at end of file

File: payment_system/templates/payment_system/invoice_detail.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ <br />
+<h1>Buyer: </h1>
+{% endblock %}
+{{company.project_name}}, <br />
+{{user.email}}
+<br />
+{% block title %}
+<p>Invoice # {{company.id}} from {{company.start_date|date:'d E Y'}}</p>
+{% endblock %}
+<table class="table" id="invoice">
```
Comment (roman-tiukh) [inline payment_system/templates/payment_system/invoice_detail.html:20]: tabs

File: payment_system/views.py [hunk 4]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
     pagination_class = None
 
 
+class InvoicePdfGenerator(generics.ListAPIView):
```
Comment (roman-tiukh) [inline payment_system/views.py:225]: RetrieveAPIView

File: payment_system/views.py [hunk 5]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
...
+    serializer_class = InvoiceSerializer
+    queryset = Invoice
+    pagination_class = None
+
+    def gener_pdf(request, pk):
```
Comment (roman-tiukh) [inline payment_system/views.py:230]:
```suggestion
    def retrieve(self, request, pk):
```

File: payment_system/views.py [hunk 6]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
     pagination_class = None
 
 
+class InvoicePdfGenerator(generics.ListAPIView):
+    serializer_class = InvoiceSerializer
+    queryset = Invoice
```
Comment (roman-tiukh) [inline payment_system/views.py:227]: Invoice its not a queryset

File: payment_system/views.py [hunk 7]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
...
+    pagination_class = None
+
+    def gener_pdf(request, pk):
+        company = get_object_or_404(Invoice.objects.all(), id=pk)
+        pdv = company.price * 0.2
+        price_with_pdv = company.price + pdv
+        user = get_object_or_404(DataOceanUser.objects.all(), id=request.user.id)
```
Comment (roman-tiukh) [inline payment_system/views.py:234]: redundant line

File: payment_system/views.py [hunk 8]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
...
+        lang = user.language
+        if lang == 'uk':
+            summ = num2words(price_with_pdv, lang='uk', to='currency', currency='UAH', separator=' ', cents=False)
+            html_string = render_to_string('payment_system/invoice_det_uk.html', {'company': company, 'summ': summ,
+                                                                                  'wpdv': price_with_pdv,
+                                                                                  'pdv': pdv, 'user': user,
```
Comment (roman-tiukh) [inline payment_system/views.py:240]: pdv -> vat

File: payment_system/views.py [hunk 9]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
...
+                                                                                  'lang': lang})
+        else:
+            if price_with_pdv == 1:
+                summ = "one hryvnia' 00 kopiyok"
```
Comment (roman-tiukh) [inline payment_system/views.py:244]:
```suggestion
                summ = "one hryvnia 00 kopiyok"
```

File: payment_system/views.py [hunk 10]
```diff
@@ -214,6 +222,49 @@ class InvoiceListView(generics.ListAPIView):
...
+        result = html.write_pdf(stylesheets=[CSS("payment_system/templates/payment_system/invoice_stylesheet.css")])
+        response = HttpResponse(content_type='application/pdf;')
+        response['Content-Disposition'] = 'inline; attachment; filename=Invoice' + str(datetime.now()) + '.pdf'
+        response['Content-Transfer-Encoding'] = 'binary'
+
+        with tempfile.NamedTemporaryFile(delete=True) as output:
+            output.write(result)
+            output.flush()
+            output = open(output.name, 'rb')
+            response.write(output.read())
```
Comment (roman-tiukh) [inline payment_system/views.py:263]: try to write this with FileResponse and io.BytesIO

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: payment_system/views.py [hunk 11]
```diff
@@ -214,6 +221,19 @@ class InvoiceListView(generics.ListAPIView):
...
+class InvoicePdfGenerator(generics.RetrieveAPIView):
+    queryset = Invoice.objects.all()
+    serializer_class = InvoiceTestSerializer
+
+    def retrieve(self, request, pk):
+        current_user = request.user
+        return self.gener_pdf(request, pk, current_user)
+
+    def gener_pdf(self, request, pk, current_user):
+        qwerty = get_object_or_404(Invoice.objects.all(), id=pk)
```
Comment (roman-tiukh) [inline payment_system/views.py:233]: qwerty?

Comment (Andrii17s) [reply payment_system/views.py:233]: Done

Comment (Andrii17s) [reply payment_system/views.py]: Done

File: payment_system/models.py [hunk 12]
```diff
@@ -333,6 +341,35 @@ def __str__(self):
     class Meta:
         ordering = ['-created_at']
 
+    @classmethod
+    def get_pdf(self, request, user):
+        seller = ''
+        with open('payment_system/static/payment_system/credentials.txt') as cr_file:
```
Comment (criticus) [inline payment_system/models.py:347]: consider moving this to settings

File: payment_system/models.py [hunk 13]
```diff
@@ -333,6 +341,35 @@ def __str__(self):
...
+
+        vat = round(request.price * 0.2, 2)
+        price_with_vat = request.price + vat
+        lang = user.language
+        with translation.override(lang):
+            if lang == 'uk':
+                summ = num2words(price_with_vat, lang='uk', to='currency', currency='UAH', separator=' ', cents=False)
+            else:
+                if price_with_vat == 1:
+                    summ = 'one hryvnia 00 kopiyok'
```
Comment (criticus) [inline payment_system/models.py:358]: better to replace with `elif` https://docs.python.org/3/tutorial/controlflow.html

File: payment_system/models.py [hunk 14]
```diff
@@ -333,6 +341,35 @@ def __str__(self):
     class Meta:
         ordering = ['-created_at']
 
+    @classmethod
+    def get_pdf(self, request, user):
```
Comment (criticus) [inline payment_system/models.py:345]: I think this method works better in `Invoice` class.  `Invoice.get_pdf()`. What do you think?

**Comment by roman-tiukh** [comment]
@criticus its not actual PR, closed.

**Pull request closed without merging**